services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java

```diff
@@ -72,116 +70,104 @@ public SAMLIdentityProviderConfig createConfig() {
-                            } else if (KeyTypes.ENCRYPTION.equals(keyDescriptorType.getUse())) {
-                                samlIdentityProviderConfig.setEncryptionPublicKey(x509KeyInfo.getTextContent());
-                            } else if (keyDescriptorType.getUse() ==  null) {
-                                defaultCertificate = x509KeyInfo.getTextContent();
-                            }
-                        }
-                    }
-
-                    if (defaultCertificate != null) {
-                        if (samlIdentityProviderConfig.getSigningCertificates().length == 0) {
-                            samlIdentityProviderConfig.addSigningCertificate(defaultCertificate);
-                        }
 
-                        if (samlIdentityProviderConfig.getEncryptionPublicKey() == null) {
-                            samlIdentityProviderConfig.setEncryptionPublicKey(defaultCertificate);
-                        }
+                    if (samlIdentityProviderConfig.getEncryptionPublicKey() == null) {
+                        samlIdentityProviderConfig.setEncryptionPublicKey(defaultCertificate);
                     }
+                }
 
-                    samlIdentityProviderConfig.setEnabledFromMetadata(entityType.getValidUntil() == null
+                samlIdentityProviderConfig.setEnabledFromMetadata(entityType.getValidUntil() == null
                         || entityType.getValidUntil().toGregorianCalendar().getTime().after(new Date(System.currentTimeMillis())));
 
-                    // check for hide on login attribute
-                    if (entityType.getExtensions() != null && entityType.getExtensions().getEntityAttributes() != null) {
-                        for (AttributeType attribute : entityType.getExtensions().getEntityAttributes().getAttribute()) {
-                            if (MACEDIR_ENTITY_CATEGORY.equals(attribute.getName())
+                // check for hide on login attribute
+                if (entityType.getExtensions() != null && entityType.getExtensions().getEntityAttributes() != null) {
+                    for (AttributeType attribute : entityType.getExtensions().getEntityAttributes().getAttribute()) {
+                        if (MACEDIR_ENTITY_CATEGORY.equals(attribute.getName())
                                 && attribute.getAttributeValue().contains(REFEDS_HIDE_FROM_DISCOVERY)) {
-                                samlIdentityProviderConfig.setHideOnLogin(true);
-                            }
+                            samlIdentityProviderConfig.setHideOnLogin(true);
                         }
-
                     }
 
-                    return samlIdentityProviderConfig.getConfig();
                 }
+
+                if (from != null) {
+                    samlIdentityProviderConfig.setEntityDescriptorUrl(from);
+                    // TODO: set useDescriptorUrl to ON by default?
```

Comment:
Is this the only reason for introducing the new method the `IdentityProviderFactory` interface? It is not clear for me from the method signature that `from` also means changing the provider config ...

Also, I'm wondering if that is not something that should not be done automatically but depend on user action.

Reply:
This method doesn't modify the provider config. It creates a new provider config to be returned to the caller. The idea is the UI calls this method, receives a config and with that config creates later a new Identity provider of this type. It is used in OIDC and SAML providers (with the well-known-address and endpoint metadata URL).

I added the `from` parameter because the URL used to retrieve the `InputStream` is not passed (and not returned in the saml metadata). In OIDC it's not needed as the well-know-address is not used to create the config, just other URLs inside the json returned are used. But I'm not modifying anything just creating a template configuration for later creation and that URL is needed.

Do you see any other way of receiving the initial URL that created the InputStream?

Reply:
If we don't need it in OIDC then it seems we need to re-think the new method you are adding.

We could avoid this by using a standard attribute name to store the location from where a configuration was loaded. It won't be SAML specific such as `entityDescriptorUrl` but a more generic name such as `configurationUrl` or something similar. I hope you get the point ...

Reply:
Done now!